crates/node/src/types/file.rs

```diff
@@ -0,0 +1,220 @@
+use crate::types::transaction;
+use crate::types::Hash;
+use eyre::Result;
+use serde::Deserialize;
+use serde::Serialize;
+use std::path::Path;
+use std::path::PathBuf;
+
+pub async fn open_task_file(
+    data_dir: &PathBuf,
+    task_id: &str,
+    path: &str,
+) -> Result<tokio::io::BufReader<tokio::fs::File>> {
+    let mut path = Path::new(path);
+    if path.is_absolute() {
+        path = path.strip_prefix("/")?;
+    }
+    let path = PathBuf::new().join(data_dir).join(task_id).join(path);
+    let fd = tokio::fs::File::open(path).await?;
+    Ok(tokio::io::BufReader::new(fd))
+}
+
+//describe file data that is stored in the database.
+// to manipulate file on disk use the equivalent type state definition File<T>
+#[derive(Clone, Debug, Default, PartialEq, Deserialize, Serialize, sqlx::FromRow)]
+pub struct DbFile {
+    pub name: String,
+    pub url: String,
+    pub checksum: Hash,
+}
+
+impl DbFile {
+    pub fn try_from_prg_data(
+        value: &transaction::ProgramData,
+    ) -> Result<Option<Self>, &'static str> {
+        let file = match value {
+            transaction::ProgramData::Input {
+                file_name,
+                file_url,
+                checksum,
+            } => Some(DbFile {
+                name: file_name.clone(),
+                url: file_url.clone(),
+                checksum: checksum.clone().into(),
+            }),
+            transaction::ProgramData::Output {
+                source_program: _,
+                file_name: _,
+            } => {
+                //Output file are not use by Tx management.
+                None
+            }
+        };
+
+        Ok(file)
+    }
+}
+
+// Type state definition of a file to manage all the different case of file manipulation.
+// Download: Use to download the file data and move them to the right place. Constructed using the other type.
+// ProofVerif: A file attached to a proof or verify Tx: Only downloaded on distant host. Nothing done on the host when the Tx has been executed.
+// Vm(Hash): File generated by the VM execution. Move from Vm to ProofVerif place.
+// Image: File attached to a Deploy Tx. Identify an image that are stored in the image directory.
+#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
+pub struct File<E> {
+    pub name: String,
+    pub url: String,
+    pub checksum: Hash,
+    pub extention: E,
+}
```

Comment:
Below I expressed my thoughts on the transaction's state domain type modeling. I think that same thinking applies on this case as well.